### ISSUE TITLE
How to use existing test results(npy)

### ISSUE THREAD

**Issue by 22xxlin** [mikel-brostrom/boxmot#1816]
### Search before asking

- [x] I have searched the Yolo Tracking [issues](https://github.com/mikel-brostrom/yolo_tracking/issues) and found no similar bug report.

### Question

My problems are very similar and https://github.com/mikel-brostrom/boxmot/issues/1156,
According to the discussion of this problem, I download the corresponding test results, https://drive.google.com/drive/folders/1zzzUROXYXt8NjxO1WUcwSzqD-nn7rPNr, I noticed this is npy format
I tried the following steps:

1. Place MOT17 into the folder named data
2. Remove all the sequences in the MOT17 that end with -DPM or -SDP. Only the ones that end with -FRCNN should remain.
3. Download detections from https://drive.google.com/drive/folders/1zzzUROXYXt8NjxO1WUcwSzqD-nn7rPNr
4. Place detections into the folder named MOT17_train_dets
5. Run python run.py --tracker strongsort
6. The results will be saved into the folder named results

But I found that the current version does not have run.py
So I tried it
python3 tracking/val.py --yolo-model yolov8n.pt --reid-model osnet_x0_25_msmt17.pt --tracking-method strongsort --verbose --source ./tracking/val_utils/data/MOT17/train --dets-file-path ./tracking/val_utils/data/MOT17/MOT17_train_dets --ci

However, according to the output information, I found that the detection information was still generated according to yolo8n, rather than using the detection information I input, namely the npy file

How should this problem be dealt with

`(Track) iusl@iusl-MS-7D04:~/D-UAV/boxmot$ python3 tracking/val.py --yolo-model yolov8n.pt --reid-model osnet_x0_25_msmt17.pt --tracking-method strongsort --verbose --source ./tracking/val_utils/data/MOT17/train --dets-file-path ./tracking/val_utils/data/MOT17/MOT17_train_dets --ci
2025-02-13 11:21:12.539 | INFO     | tracking.utils:download_mot_eval_tools:118 - Evaluation repo already downloaded or an error occurred: Cmd('git') failed due to: exit code(128)
  cmdline: git clone -v -- https://github.com/JonathonLuiten/TrackEval /home/iusl/D-UAV/boxmot/tracking/val_utils
  stderr: 'fatal: destination path '/home/iusl/D-UAV/boxmot/tracking/val_utils' already exists and is not an empty directory.
'
2025-02-13 11:21:13.694 | INFO     | tracking.utils:download_mot_dataset:170 - MOT17.zip is already fully downloaded.
2025-02-13 11:21:13.695 | INFO     | tracking.utils:unzip_mot_dataset:253 - MOT17 folder already exists.
Sequence MOT17-09 is already cleaned. Skipping.
Sequence MOT17-10 is already cleaned. Skipping.
Sequence MOT17-13 is already cleaned. Skipping.
Sequence MOT17-11 is already cleaned. Skipping.
Sequence MOT17-04 is already cleaned. Skipping.
Sequence MOT17-05 is already cleaned. Skipping.
Sequence MOT17-02 is already cleaned. Skipping.
MOT17 Cleanup completed!
2025-02-13 11:21:13.696 | INFO     | __main__:run_generate_dets_embs:383 - Generating detections and embeddings for data under tracking/val_utils/data/MOT17/train/MOT17-02 [1/7 seqs]
2025-02-13 11:21:13.916 | INFO     | boxmot.utils.torch_utils:select_device:52 - Yolo Tracking v11.0.6 🚀 Python-3.9.0 torch-2.2.2+cu121
CUDA:0 (NVIDIA GeForce RTX 3070, 7940MiB)
2025-02-13 11:21:13.951 | SUCCESS  | boxmot.appearance.reid_model_factory:load_pretrained_weights:183 - Loaded pretrained weights from osnet_x0_25_msmt17.pt
Frames: 600it [00:13, 46.02it/s]
2025-02-13 11:21:27.003 | INFO     | __main__:run_generate_dets_embs:383 - Generating detections and embeddings for data under tracking/val_utils/data/MOT17/train/MOT17-04 [2/7 seqs]`

**Comment by 22xxlin** [comment]
After a day's struggle finally solved this problem!!


First of all, npy files do generate the dets and embs folders that your toolkit needs from your scripts, but the one thing to note is that the generated files seem to be missing the first line, and I'm not sure if it's the author script or because I changed it later
In short, a missing first line results in misaligned columns and no file to be found
![Image](https://github.com/user-attachments/assets/268b7e22-4e2a-49da-b2bf-fd1156e96923)
Therefore, we can generate the dets and embs we need through the author's script and corresponding npy files, or through the author's download connection to run files, which is the same information

> For reproducing the old results you can set --conf 0.2. For reproducing the new results, download the detections and emebedding + second half of MOT17 from here:

https://github.com/mikel-brostrom/boxmot/releases/tag/v10.0.83

or use this script by aythor:
`import os
import cv2
import numpy as np
from pathlib import Path
import argparse


base_dirs_with_thresholds = [
    ("/Users/mikel.brostrom/boxmot/tracking/val_utils/data/MOT17/train/MOT17-02", 302),
    ("/Users/mikel.brostrom/boxmot/tracking/val_utils/data/MOT17/train/MOT17-04", 527),
    ("/Users/mikel.brostrom/boxmot/tracking/val_utils/data/MOT17/train/MOT17-05", 420),
    ("/Users/mikel.brostrom/boxmot/tracking/val_utils/data/MOT17/train/MOT17-09", 264),
    ("/Users/mikel.brostrom/boxmot/tracking/val_utils/data/MOT17/train/MOT17-10", 329),
    ("/Users/mikel.brostrom/boxmot/tracking/val_utils/data/MOT17/train/MOT17-11", 452),
    ("/Users/mikel.brostrom/boxmot/tracking/val_utils/data/MOT17/train/MOT17-13", 377)
]


def get_seq_paths():
    # Define the exact paths for each sequence
    seq_paths = {
        "MOT17-13": "/Users/mikel.brostrom/boxmot/tracking/val_utils/data/MOT17/train/MOT17-13/img1",
        "MOT17-11": "/Users/mikel.brostrom/boxmot/tracking/val_utils/data/MOT17/train/MOT17-11/img1",
        "MOT17-10": "/Users/mikel.brostrom/boxmot/tracking/val_utils/data/MOT17/train/MOT17-10/img1",
        "MOT17-09": "/Users/mikel.brostrom/boxmot/tracking/val_utils/data/MOT17/train/MOT17-09/img1",
        "MOT17-05": "/Users/mikel.brostrom/boxmot/tracking/val_utils/data/MOT17/train/MOT17-05/img1",
        "MOT17-04": "/Users/mikel.brostrom/boxmot/tracking/val_utils/data/MOT17/train/MOT17-04/img1",
        "MOT17-02": "/Users/mikel.brostrom/boxmot/tracking/val_utils/data/MOT17/train/MOT17-02/img1"
    }
    
    imgs = {}
    for seq, path in seq_paths.items():
        img_files = [os.path.join(path, file) for file in os.listdir(path) if file.endswith(".jpg")]
        img_files.sort(key=lambda x: int(x.split("/")[-1].split(".")[0]))  # Sorting based on file name
        imgs[seq] = img_files
    return imgs, list(seq_paths.keys())


# Get the threshold for the sequence
def get_threshold_for_sequence(seq):
    for base_dir, threshold in base_dirs_with_thresholds:
        if seq in base_dir:
            return threshold
    return 0  # Default threshold if not found (should not happen)


def parse_options():
    parser = argparse.ArgumentParser()
    parser.add_argument('--reid', type=str, default="osnet_x1_0_dukemtmcreid.pt", help='model.pt path')
    parser.add_argument('--dataset', type=str, default='train', help='dataset type')
    parser.add_argument('--device', type=str, default='cpu', help='device \'cpu\' or \'0\', \'1\', ... for gpu')
    parser.add_argument('--ot', type=float, default=0.2)
    return parser.parse_args()

def process_detection(row, frame_no):
    tlwh = row[2:6]
    # Include the frame number as the first element
    return np.array([frame_no, tlwh[0], tlwh[1], tlwh[0] + tlwh[2], tlwh[1] + tlwh[3], row[6], 0])

def save_to_txt(file_path, data):
    with open(file_path, 'a') as f:
        for item in data:
            f.write(' '.join(map(str, item)) + '\n')

if __name__ == "__main__":
    args = parse_options()
    
    # Get image paths for each sequence
    imgs, seq_names = get_seq_paths()

    total_dets = 0
    total_frames = 0

    # Define paths for output
    output_dir_dets = Path("runs/dets_n_embs/yolov8x/dets")
    output_dir_embs = Path("runs/dets_n_embs/yolov8x/embs/osnet_x1_0_dukemtmcreid")
    output_dir_dets.mkdir(parents=True, exist_ok=True)
    output_dir_embs.mkdir(parents=True, exist_ok=True)

    for seq in seq_names:
        print(f"Processing sequence: {seq}")
        seq_imgs = imgs[seq]
        
        # Get the threshold for the current sequence
        offset = get_threshold_for_sequence(seq)

        # Define text file paths for detections and embeddings for this sequence
        dets_txt_path = output_dir_dets / f"{seq}.txt"
        embs_txt_path = output_dir_embs / f"{seq}.txt"

        # Assuming detections for each sequence are stored in .npy files
        det_file = Path(f"./npy_files/{seq}-FRCNN.npy")
        seq_det = np.load(det_file, allow_pickle=True)
        print(f"Loaded .npy file for {seq}, shape: {seq_det.shape}")

        for frame_no, img_path in enumerate(seq_imgs):
            frame = cv2.imread(img_path)
            frame_dets = seq_det[seq_det[:, 0] == frame_no + offset]
            
            print(f"Detections for frame {frame_no + 1}: {len(frame_dets)}")

            if len(frame_dets) < 1:
                continue

            total_dets += len(frame_dets)
            total_frames += 1

            processed_dets = np.array([process_detection(row, frame_no + 1) for row in frame_dets])
            features = np.array([row[10:] for row in frame_dets])  # Assuming embeddings are the last columns

            # Save detections and embeddings to their respective text files
            save_to_txt(dets_txt_path, processed_dets)
            save_to_txt(embs_txt_path, features)

        print(f"Processed {len(seq_imgs)} frames for sequence {seq}")

    print(f"Total frames processed: {total_frames}")
    print(f"Total detections: {total_dets}")
`

Second, after I compared the same files, I found that the MOTA I generated was negative

This is because I used MOT17 instead of MOT17-50, and I looked at the related problem discussion and it seems to use the first and second half of the data set, which may have something to do with gt files and image files
In short, when I used MOT17-50, I reproduced the accuracy described by the author
`Starting tracking on:
	tracking/val_utils/data/MOT17-50/train/MOT17-04/img1
with preloaded dets
	(runs/dets_n_embs/yolov8x/dets/MOT17-04.txt)
and embs
	(runs/dets_n_embs/yolov8x/embs/osnet_x1_0_dukemtmcreid/MOT17-04.txt)
using
	bytetrack
Frames: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1050/1050 [00:07<00:00, 143.29it/s]

Eval Config:
USE_PARALLEL         : True                          
NUM_PARALLEL_CORES   : 4                             
BREAK_ON_ERROR       : True                          
RETURN_ON_ERROR      : False                         
LOG_ON_ERROR         : /home/iusl/D-UAV/boxmot/tracking/val_utils/error_log.txt
PRINT_RESULTS        : True                          
PRINT_ONLY_COMBINED  : False                         
PRINT_CONFIG         : True                          
TIME_PROGRESS        : True                          
DISPLAY_LESS_PROGRESS : False                         
OUTPUT_SUMMARY       : True                          
OUTPUT_EMPTY_CLASSES : True                          
OUTPUT_DETAILED      : True                          
PLOT_CURVES          : True                          

MotChallenge2DBox Config:
PRINT_CONFIG         : True                          
GT_FOLDER            : /home/iusl/D-UAV/boxmot/tracking/val_utils/data/MOT17/train
TRACKERS_FOLDER      : /home/iusl/D-UAV/boxmot/runs/mot/yolov8x_osnet_x1_0_dukemtmcreid_bytetrack
OUTPUT_FOLDER        : None                          
TRACKERS_TO_EVAL     : ['']                          
CLASSES_TO_EVAL      : ['pedestrian']                
BENCHMARK            :                               
SPLIT_TO_EVAL        : train                         
INPUT_AS_ZIP         : False                         
DO_PREPROC           : True                          
TRACKER_SUB_FOLDER   :                               
OUTPUT_SUB_FOLDER    :                               
TRACKER_DISPLAY_NAMES : None                          
SEQMAP_FOLDER        : None                          
SEQMAP_FILE          : None                          
SEQ_INFO             : {'MOT17-02': None, 'MOT17-04': None, 'MOT17-09': None, 'MOT17-13': None, 'MOT17-10': None, 'MOT17-05': None, 'MOT17-11': None}
GT_LOC_FORMAT        : {gt_folder}/{seq}/gt/gt.txt   
SKIP_SPLIT_FOL       : True                          

CLEAR Config:
METRICS              : ['HOTA', 'CLEAR', 'Identity'] 
THRESHOLD            : 0.5                           
PRINT_CONFIG         : True                          

Identity Config:
METRICS              : ['HOTA', 'CLEAR', 'Identity'] 
THRESHOLD            : 0.5                           
PRINT_CONFIG         : True                          

Evaluating 1 tracker(s) on 7 sequence(s) for 1 class(es) on MotChallenge2DBox dataset using the following metrics: HOTA, CLEAR, Identity, Count


Evaluating 


All sequences for  finished in 0.66 seconds

HOTA: -pedestrian                  HOTA      DetA      AssA      DetRe     DetPr     AssRe     AssPr     LocA      OWTA      HOTA(0)   LocA(0)   HOTALocA(0)
MOT17-02                           7.5767    7.373     8.2033    8.8443    25.097    9.9853    33.544    66.172    8.3383    16.355    38.976    6.3744    
MOT17-04                           16.524    11.501    26.012    14.818    30.258    28.619    60.49     73.798    18.894    25.351    50.094    12.699    
MOT17-05                           7.1691    6.4076    8.3438    7.9294    19.923    14.523    19.062    64.1      8.0125    19.792    32.135    6.3601    
MOT17-09                           4.3364    4.2278    4.8398    5.5528    12.701    6.4799    16.737    63.6      4.9907    12.741    30.234    3.8521    
MOT17-10                           1.8966    2.1257    1.8984    2.6929    7.7242    3.4267    5.0654    60.398    2.1447    7.2766    21.04     1.531     
MOT17-11                           7.4058    6.9347    8.0835    8.6466    20.454    11.784    22.365    63.303    8.2921    19.708    35.21     6.9392    
MOT17-13                           1.0992    0.75853   1.8594    0.90778   4.1592    2.9963    10.989    62.28     1.204     3.3228    27.126    0.90133   
COMBINED                           11.827    7.6201    21.646    9.6189    23.521    24.591    52.687    70.659    13.399    19.713    42.496    8.3771    

CLEAR: -pedestrian                 MOTA      MOTP      MODA      CLR_Re    CLR_Pr    MTR       PTR       MLR       sMOTA     CLR_TP    CLR_FN    CLR_FP    IDSW      MT        PT        ML        Frag      
MOT17-02                           -18.557   64.539    -18.105   8.5679    24.313    0         11.29     88.71     -21.595   1592      16989     4956      84        0         7         55        147       
MOT17-04                           -18.346   73.778    -18.182   15.394    31.435    0         21.687    78.313    -22.383   7321      40236     15968     78        0         18        65        104       
MOT17-05                           -28.64    63.732    -28.235   5.7829    14.53     0         6.015     93.985    -30.737   400       6517      2353      28        0         8         125       41        
MOT17-09                           -37.014   64.221    -36.808   3.4554    7.9038    0         0         100       -38.25    184       5141      2144      11        0         0         26        8         
MOT17-10                           -33.235   59.707    -33.18    0.84119   2.4129    0         0         100       -33.574   108       12731     4368      7         0         0         57        11        
MOT17-11                           -29.822   61.313    -29.536   6.3692    15.066    0         2.6667    97.333    -32.286   601       8835      3388      27        0         2         73        45        
MOT17-13                           -21.053   62.585    -20.898   0.46384   2.1251    0         0.90909   99.091    -21.227   54        11588     2487      18        0         1         109       13        
COMBINED                           -22.847   70.844    -22.622   9.1365    22.341    0         6.5934    93.407    -25.511   10260     102037    35664     253       0         36        510       369       

Identity: -pedestrian              IDF1      IDR       IDP       IDTP      IDFN      IDFP      
MOT17-02                           7.8157    5.285     14.997    982       17599     5566      
MOT17-04                           16.645    12.398    25.317    5896      41661     17393     
MOT17-05                           4.6536    3.2529    8.1729    225       6692      2528      
MOT17-09                           2.4043    1.7277    3.9519    92        5233      2236      
MOT17-10                           0.66994   0.45175   1.2958    58        12781     4418      
MOT17-11                           4.7374    3.3701    7.9719    318       9118      3671      
MOT17-13                           0.26793   0.1632    0.74774   19        11623     2522      
COMBINED                           9.5942    6.7589    16.527    7590      104707    38334     

Count: -pedestrian                 Dets      GT_Dets   IDs       GT_IDs    
MOT17-02                           6548      18581     88        62        
MOT17-04                           23289     47557     108       83        
MOT17-05                           2753      6917      59        133       
MOT17-09                           2328      5325      23        26        
MOT17-10                           4476      12839     59        57        
MOT17-11                           3989      9436      70        75        
MOT17-13                           2541      11642     48        110       
COMBINED                           45924     112297    455       546       

{"HOTA": 11.827, "MOTA": -22.847, "IDF1": 9.5942}
(Track) iusl@iusl-MS-7D04:~/D-UAV/boxmot$ n`

In general, we want to make sure first,
Whether dets and embs files are pre-trained in this data set, from the perspective of detection, the effect generated directly from a general YOL.PT that has not been trained in this data set is definitely much worse than the effect after training on this data set, so it is recommended to determine the pre-training weight of the detection. Generate more accurate dets and embs results
Second, if you reproduce the accuracy of MOT17, make sure you use MOT17-50

I hope I'm right
I'm exhausted today

**Comment by mikel-brostrom** [comment]
The easiest way is still to just download `runs.zip`
